### DISCUSSION TITLE
got unspecified error when translating c++ headers

### DISCUSSION THREAD

**Discussion by lgc2333** [swig/swig#2388] [Q&A] [answered]
This is a part of my workspace

<img width="123" alt="image" src="https://user-images.githubusercontent.com/59048777/193927181-27b9ade5-be29-4a61-abfc-d5245da0a8e9.png">

submodule LLSDK here -> https://github.com/LiteLDev/LiteLoaderSDK

file `logger.i`
```swig
%module ll

%{
#include "../LLSDK/include/llapi/LoggerAPI.h"
%}

%include "../LLSDK/include/llapi/LoggerAPI.h"
```

the command i run and the error i got
```
C:\Users\Administrator\Desktop\stu-files\go\TestLLGo> swig -c++ -go -intgosize 64 src/ll/logger.i
src\ll\..\LLSDK\include\llapi\LoggerAPI.h(66) : Error: Syntax error in input(3).
```

when i removed keyword `explicit` in `src\LLSDK\include\llapi\LoggerAPI.h` at line 66 and 79, then rerun the command, it stucks there.

<img width="591" alt="image" src="https://user-images.githubusercontent.com/59048777/193928747-003506dd-e52d-4869-bda5-f9fb76779e78.png">
<img width="374" alt="image" src="https://user-images.githubusercontent.com/59048777/193929358-a7c4bef8-8e90-4d91-82b7-e3595e5c6ac8.png">

i'm using swig 4.0.2 on windows. i don't know whether this is a bug or a problem with the SWIG interface file, so i created a discussion, not a issue. hope someone can help me!
(sorry about my english level)

**Comment by ojwb** [comment, marked as answer]
The syntax error is due to the `LIAPI` not the `explicit` (removing the `explict` allows it to parse because then `LIAPI` looks like an undeclared type).  SWIG doesn't follow `#include` in a `%include`-d file by default.  The simplest fix here is to define `LIAPI` as empty:

```
%module ll

%{
#include "../LLSDK/include/llapi/LoggerAPI.h"
%}

#define LIAPI
%include "../LLSDK/include/llapi/LoggerAPI.h"
```

With that fix, the command also hangs for me with SWIG 4.1.0, but with 4.1.1 it completes, so the answer there is to use the latest release.  I didn't try to narrow down the exact change that fixed this, but from the changelog I suspect it's:

```
2022-11-02: wsfulton
            #2418 Fix infinite loop handling non-type template parameters.

            Fixes infinite loop due to () brackets in a non-type template
            parameter containing an expression.
```
